### PR TITLE
drivers: watchdog: Make SOC reset as an option in counter watchdog

### PR DESCRIPTION
**Pull request by duda-patryk**
When using watchdog which doesn't support callbacks (e.g. STM32 IWDG) it's very useful to have counter based watchdog with shorter timeout. This way we can gather useful information before IWDG resets SOC.

In above case it's usually not desired to reset SOC when counter watchdog is triggerd, because we can lose information about reset cause. For example, STM32 can determine if the reset was caused by watchdog.

Another use case for counter watchdog without SOC reset is to detect slow code paths.

Best regards,
Patryk

### CODE REVIEW

**Comment by semihalf-duda-patryk** [comment]
Added commit with simple implementation of free counter channels tracking. This allows to install watchdog more than number of counter channels.

File: drivers/watchdog/wdt_counter.c [hunk 1]
```diff
@@ -56,6 +57,7 @@ static void counter_alarm_callback(const struct device *dev,
 	struct wdt_counter_data *data = wdt_dev->data;
 
 	counter_stop(dev);
+	data->used[chan_id] = false;
```
Comment (martinjaeger) [inline drivers/watchdog/wdt_counter.c:60]: The assumption that a channel can be freed as soon as the callback was triggered is not documented anywhere. Is it the same across other watchdog drivers? I don't think so.

Comment (semihalf-duda-patryk) [reply drivers/watchdog/wdt_counter.c:60]: You are right. Most drivers don't support watchdog channels. NRFX watchdog driver doesn't reuse channels.

Would you accept patch which makes SOC reset as an option https://github.com/zephyrproject-rtos/zephyr/pull/53195/commits/8b70d91bef88284462e2be69efbfc5463dc33bbd. If yes, I will remove free channels tracking from this PR.

Comment (martinjaeger) [reply drivers/watchdog/wdt_counter.c:60]: Yes, that sounds good. Let's keep the channel tracking out for now.

Comment (semihalf-duda-patryk) [reply drivers/watchdog/wdt_counter.c]: Done! Please check if it's good now.

**Comment by semihalf-duda-patryk** [comment]
Friendly ping

**Review by martinjaeger — APPROVED** [review]

**Review by nordic-krch — APPROVED** [review]

**Pull request merged**